<!doctype html>
<html class="no-js" lang="en">

<head>
  <!-- Global site tag (gtag.js) - Google Analytics -->
  <script async src="https://www.googletagmanager.com/gtag/js?id=UA-134931760-1"></script>
  <script>
    window.dataLayer = window.dataLayer || [];
    function gtag(){dataLayer.push(arguments);}
    gtag('js', new Date());

    gtag('config', 'UA-134931760-1');
  </script>

  <meta charset="utf-8">
  <meta http-equiv="x-ua-compatible" content="ie=edge">
  <title>Fix 7 common exception handling errors in java</title>
  <link rel="icon" type="image/x-icon" href="/favicon.ico" />
  <meta name="description" content="Hi, Habr! I present to you the translation of the article Fixing 7 Common Java Exception Handling Mistakes by Thorben Janssen. 

 Handling an exceptio...">
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">

  <link rel="stylesheet" href="../../css/main.css">

  <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
  <script>window.jQuery || document.write('<script src="../../js/vendors/jquery-3.3.1.min.js"><\/script>')</script>

  <script async src="//pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script>
  <script>
       (adsbygoogle = window.adsbygoogle || []).push({
            google_ad_client: "ca-pub-6974184241884155",
            enable_page_level_ads: true
       });
  </script>
</head>

<body>
  <!--[if lte IE 9]>
    <p class="browserupgrade">You are using an <strong>outdated</strong> browser. Please <a href="https://browsehappy.com/">upgrade your browser</a> to improve your experience and security.</p>
  <![endif]-->
  <header class="page-header js-page-header">
    <a class="page-header-logo-container" href="https://weekly-geekly.github.io/index.html"></a>
    <div class="page-header-text">Geekly Articles each Day</div>
  </header>
  <nav class="page-headings-container js-page-headings-container"></nav>
  <div class="tools-bar js-tools-bar">
    <!-- <a href="../../search.html" title="Search">üîé</a> -->
    <a class="js-list-of-headings-button" data-state="closed" href="#" title="Headings">üìú</a>
    <a class="js-go-to-top-button" href="#" title="Go to Top">‚¨ÜÔ∏è</a>
    <a class="js-go-to-bottom-button" href="#" title="Go to Bottom">‚¨áÔ∏è</a>
  </div>
  <a href="http://bit.ly/donateToWeeklyGeekly" class="donate-btn">DONATE</a>
  <section class="page js-page"><h1>Fix 7 common exception handling errors in java</h1><div class="post__text post__text-html js-mediator-article">  Hi, Habr!  I present to you the translation of the article <a href="https://dzone.com/articles/fixing-7-common-java-exception-handling-mistakes">Fixing 7 Common Java Exception Handling Mistakes</a> by Thorben Janssen. <br><br>  Handling an exception is one of the most common, but not necessarily one of the simplest tasks.  This is still one of the frequently discussed topics in experienced teams, and there are several advanced techniques and common mistakes you should be aware of. <br><br>  Here are a few things to avoid when handling exceptions in your application. <br><a name="habracut"></a><br><h3>  Error 1: declaration java.lang.Exception or java.lang.Throwable </h3><br>  As you already know, you need to either declare or handle the checked exception.  But checked exceptions are not the only ones you can specify.  You can use any subclass of java.lang.Throwable in the throws clause.  Thus, instead of specifying two different exceptions thrown by the following code snippet, you can simply use the java.lang.Exception exception in the throws clause. 
      <br>
        <script async src="//pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script>
        <ins class="adsbygoogle"
          style="display:block; text-align:center;"
          data-ad-layout="in-article"
          data-ad-format="fluid"
          data-ad-client="ca-pub-6974184241884155"
          data-ad-slot="8945601208"></ins>
        <script>
          (adsbygoogle = window.adsbygoogle || []).push({});
        </script>
      <br>
    <pre><code class="java hljs"><span class="hljs-function"><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">public</span></span></span><span class="hljs-function"> </span><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">void</span></span></span><span class="hljs-function"> </span><span class="hljs-title"><span class="hljs-function"><span class="hljs-title">doNotSpecifyException</span></span></span><span class="hljs-params"><span class="hljs-function"><span class="hljs-params">()</span></span></span><span class="hljs-function"> </span><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">throws</span></span></span><span class="hljs-function"> Exception </span></span>{ doSomething(); } <span class="hljs-function"><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">public</span></span></span><span class="hljs-function"> </span><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">void</span></span></span><span class="hljs-function"> </span><span class="hljs-title"><span class="hljs-function"><span class="hljs-title">doSomething</span></span></span><span class="hljs-params"><span class="hljs-function"><span class="hljs-params">()</span></span></span><span class="hljs-function"> </span><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">throws</span></span></span><span class="hljs-function"> NumberFormatException, IllegalArgumentException </span></span>{ <span class="hljs-comment"><span class="hljs-comment">// do something }</span></span></code> </pre> <br>  But this does not mean that you should do this.  Specifying Exeption or Throwable makes it almost impossible to properly handle them when calling your method. The only information that the method that you are calling receives is that something can go wrong.  But you do not share any information about any exceptional events that may occur.  You hide this information for generalized reasons for throwing exceptions. It becomes even worse when your application changes over time.  Throwing out generalized exceptions hides all exception changes that the caller must expect and handle.  This can lead to several unexpected errors that need to be found in the test case instead of a compiler error. <br><br><h4>  Use specific classes </h4><br>  It is much better to specify the most specific exception classes, even if you have to use several of them.  This tells the calling device which exceptional events to handle.  It also allows you to update the throw clause when your method throws an additional exception.  This way, your clients are aware of the changes and even get an error if you change the exceptions thrown.  Such an exception is much easier to find and handle than an exception that appears only when you run a specific test case. <br><br><pre> <code class="java hljs"><span class="hljs-function"><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">public</span></span></span><span class="hljs-function"> </span><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">void</span></span></span><span class="hljs-function"> </span><span class="hljs-title"><span class="hljs-function"><span class="hljs-title">specifySpecificExceptions</span></span></span><span class="hljs-params"><span class="hljs-function"><span class="hljs-params">()</span></span></span><span class="hljs-function"> </span><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">throws</span></span></span><span class="hljs-function"> NumberFormatException, IllegalArgumentException </span></span>{ doSomething(); }</code> </pre> <br><h3>  Error 2: interception of generalized exceptions </h3><br>  The severity of this error depends on which software component you are implementing and where you find the exception.  It might be nice to catch a java.lang.Exception in the main method of your Java SE application.  But you should prefer to catch certain exceptions if you are implementing a library or working on deeper layers of your application. <br><br>  This has several advantages.  This approach allows you to handle each exception class differently and does not allow you to catch exceptions that you did not expect. <br><br>  But keep in mind that the first catch block that handles the exception class or one of its superclasses will catch it.  Therefore, be sure to catch the most specific class first.  Otherwise, your IDE will show an error message or a warning about an unreachable block of code. <br><br><pre> <code class="java hljs"><span class="hljs-keyword"><span class="hljs-keyword">try</span></span> { doSomething(); } <span class="hljs-keyword"><span class="hljs-keyword">catch</span></span> (NumberFormatException e) { <span class="hljs-comment"><span class="hljs-comment">// handle the NumberFormatException log.error(e); } catch (IllegalArgumentException e) { // handle the IllegalArgumentException log.error(e); }</span></span></code> </pre> <br><h3>  Error 3: Logging and Forwarding Exceptions </h3><br>  This is one of the most popular errors when handling Java exceptions.  It may seem logical to register an exception where it was thrown, and then forward it to the caller, which can implement specific processing for a particular use case.  But you should not do this for three reasons: <br><br>  1. You do not have enough information about the use case that the caller of your method wants to implement.  An exception may be part of the expected behavior and handled by the client.  In this case, there is no need to register it.  This will add a false error message to the log file, which should be filtered by your operations team. <br><br>  2. The journal message does not provide any information that is not yet part of the exception itself.  Its trace and stack trace should contain all the necessary information about an exceptional event.  The message describes this, and the stack trace contains detailed information about the class, method, and line in which it occurred. <br><br>  3. You can register the same exception several times when you register it in every catch block that catches it.  This will ruin the statistics in your monitoring tool and make it difficult to read the log file for your operations and the development team. <br><br><h4>  Log the exception where you handle it. </h4><br>  Thus, it is best to register an exception when you handle it.  As in the following code snippet.  The doSomething method throws an exception.  The doMore method simply indicates it, because the developer does not have enough information to process it.  It is then processed in the doEvenMore method, which also writes a log message. <br><br><pre> <code class="java hljs"><span class="hljs-function"><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">public</span></span></span><span class="hljs-function"> </span><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">void</span></span></span><span class="hljs-function"> </span><span class="hljs-title"><span class="hljs-function"><span class="hljs-title">doEvenMore</span></span></span><span class="hljs-params"><span class="hljs-function"><span class="hljs-params">()</span></span></span><span class="hljs-function"> </span></span>{ <span class="hljs-keyword"><span class="hljs-keyword">try</span></span> { doMore(); } <span class="hljs-keyword"><span class="hljs-keyword">catch</span></span> (NumberFormatException e) { <span class="hljs-comment"><span class="hljs-comment">// handle the NumberFormatException } catch (IllegalArgumentException e) { // handle the IllegalArgumentException } } public void doMore() throws NumberFormatException, IllegalArgumentException { doSomething(); } public void doSomething() throws NumberFormatException, IllegalArgumentException { // do something }</span></span></code> </pre> <br><h3>  Error 4: Using Exceptions to Control Flow </h3><br>  Using exceptions to control the flow of your application is considered anti-pattern for two main reasons: <br><br>  They basically work as a Go To statement, because they cancel the execution of the code block and go to the first catch block that handles the exception.  This makes the code very difficult to read. <br><br>  They are not as effective as general Java control structures.  As the name implies, you should only use them for exceptional events, and the JVM does not optimize them in the same way as other code. Thus, it is better to use the right conditions to break your loops or if-else instructions to decide which blocks code must be executed. <br><br><h3>  Error 5: remove the cause of the exception </h3><br>  Sometimes you may need to wrap one exception into another.  Your team may have decided to use a special business exception with error codes and single processing.  There is nothing wrong with this approach if you do not eliminate the cause. <br><br>  When you create a new exception, you should always set the initial exception as the reason.  Otherwise, you lose track of the message and the stack, which describe the exceptional event that caused your exception.  The Exception class and all its subclasses provide several constructor methods that take the original exception as a parameter and set it as the cause. <br><br><pre> <code class="java hljs"><span class="hljs-keyword"><span class="hljs-keyword">try</span></span> { doSomething(); } <span class="hljs-keyword"><span class="hljs-keyword">catch</span></span> (NumberFormatException e) { <span class="hljs-keyword"><span class="hljs-keyword">throw</span></span> <span class="hljs-keyword"><span class="hljs-keyword">new</span></span> MyBusinessException(e, ErrorCode.CONFIGURATION_ERROR); } <span class="hljs-keyword"><span class="hljs-keyword">catch</span></span> (IllegalArgumentException e) { <span class="hljs-keyword"><span class="hljs-keyword">throw</span></span> <span class="hljs-keyword"><span class="hljs-keyword">new</span></span> MyBusinessException(e, ErrorCode.UNEXPECTED); }</code> </pre> <br><h3>  Error 6: Generalization of Exceptions </h3><br>  When you generalize an exception, you catch a specific, for example, NumberFormatException, and instead generate a nonspecific java.lang.Exception.  This is similar, but even worse than the first error I described in this article.  It not only hides information about a specific error case on your API, but also makes access difficult. <br><br><pre> <code class="java hljs"><span class="hljs-function"><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">public</span></span></span><span class="hljs-function"> </span><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">void</span></span></span><span class="hljs-function"> </span><span class="hljs-title"><span class="hljs-function"><span class="hljs-title">doNotGeneralizeException</span></span></span><span class="hljs-params"><span class="hljs-function"><span class="hljs-params">()</span></span></span><span class="hljs-function"> </span><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">throws</span></span></span><span class="hljs-function"> Exception </span></span>{ <span class="hljs-keyword"><span class="hljs-keyword">try</span></span> { doSomething(); } <span class="hljs-keyword"><span class="hljs-keyword">catch</span></span> (NumberFormatException e) { <span class="hljs-keyword"><span class="hljs-keyword">throw</span></span> <span class="hljs-keyword"><span class="hljs-keyword">new</span></span> Exception(e); } <span class="hljs-keyword"><span class="hljs-keyword">catch</span></span> (IllegalArgumentException e) { <span class="hljs-keyword"><span class="hljs-keyword">throw</span></span> <span class="hljs-keyword"><span class="hljs-keyword">new</span></span> Exception(e); } }</code> </pre> <br>  As you can see in the following code snippet, even if you know what exceptions a method can cause, you can't just catch them.  You need to catch the generic Exception class and then check the type of its cause.  This code is not only cumbersome to implement, but also difficult to read.  It gets even worse if you combine this approach with error 5. This removes all the information about the exceptional event. <br><br><pre> <code class="java hljs"><span class="hljs-keyword"><span class="hljs-keyword">try</span></span> { doNotGeneralizeException(); } <span class="hljs-keyword"><span class="hljs-keyword">catch</span></span> (Exception e) { <span class="hljs-keyword"><span class="hljs-keyword">if</span></span> (e.getCause() <span class="hljs-keyword"><span class="hljs-keyword">instanceof</span></span> NumberFormatException) { log.error(<span class="hljs-string"><span class="hljs-string">"NumberFormatException: "</span></span> + e); } <span class="hljs-keyword"><span class="hljs-keyword">else</span></span> <span class="hljs-keyword"><span class="hljs-keyword">if</span></span> (e.getCause() <span class="hljs-keyword"><span class="hljs-keyword">instanceof</span></span> IllegalArgumentException) { log.error(<span class="hljs-string"><span class="hljs-string">"IllegalArgumentException: "</span></span> + e); } <span class="hljs-keyword"><span class="hljs-keyword">else</span></span> { log.error(<span class="hljs-string"><span class="hljs-string">"Unexpected exception: "</span></span> + e); } }</code> </pre> <br>  So which is the best approach? <br><br><h4>  Be specific and save the reason for the exception. </h4><br>  The exceptions that you throw should always be as specific as possible.  And if you wrap an exception, you must also set the original exception as the reason not to lose the stack trace and other information describing the exceptional event. <br><br><pre> <code class="java hljs"><span class="hljs-keyword"><span class="hljs-keyword">try</span></span> { doSomething(); } <span class="hljs-keyword"><span class="hljs-keyword">catch</span></span> (NumberFormatException e) { <span class="hljs-keyword"><span class="hljs-keyword">throw</span></span> <span class="hljs-keyword"><span class="hljs-keyword">new</span></span> MyBusinessException(e, ErrorCode.CONFIGURATION_ERROR); } <span class="hljs-keyword"><span class="hljs-keyword">catch</span></span> (IllegalArgumentException e) { <span class="hljs-keyword"><span class="hljs-keyword">throw</span></span> <span class="hljs-keyword"><span class="hljs-keyword">new</span></span> MyBusinessException(e, ErrorCode.UNEXPECTED); }</code> </pre> <br><h3>  Error 7: adding unnecessary exception conversions </h3><br>  As I explained earlier, it can be useful to wrap the exceptions in custom ones, if you set the original exception as the reason.  But some architects overdo it and introduce a special exception class for each architectural level.  Thus, they catch an exception in the level of persistence and transfer it to MyPersistenceException.  The business layer catches and wraps it in a MyBusinessException, and it continues until it reaches the API level or is processed. <br><br><pre> <code class="java hljs"><span class="hljs-function"><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">public</span></span></span><span class="hljs-function"> </span><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">void</span></span></span><span class="hljs-function"> </span><span class="hljs-title"><span class="hljs-function"><span class="hljs-title">persistCustomer</span></span></span><span class="hljs-params"><span class="hljs-function"><span class="hljs-params">(Customer c)</span></span></span><span class="hljs-function"> </span><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">throws</span></span></span><span class="hljs-function"> MyPersistenceException </span></span>{ <span class="hljs-comment"><span class="hljs-comment">// persist a Customer } public void manageCustomer(Customer c) throws MyBusinessException { // manage a Customer try { persistCustomer(c); } catch (MyPersistenceException e) { throw new MyBusinessException(e, e.getCode()); } } public void createCustomer(Customer c) throws MyApiException { // create a Customer try { manageCustomer(c); } catch (MyBusinessException e) { throw new MyApiException(e, e.getCode()); } }</span></span></code> </pre> <br>  It is easy to see that these additional exception classes offer no advantages.  They simply introduce additional layers that wrap the exception.  And although it would be fun to wrap a gift in a variety of colorful paper, this is not a very good approach to software development. <br><br><h4>  Be sure to add information </h4><br>  Just think about the code that should handle the exception or about yourself when you need to find the problem that caused the exception.  First you need to break through several levels of exceptions to find the root cause.  And until today, I have never seen the application that used this approach, and added useful information with each layer of the exception.  They either summarize the error message and code, or provide redundant information. <br><br>  Therefore, be careful with the number of custom exception classes that you enter.  You should always ask yourself if the new exception class provides additional information or other benefits.  In most cases, to achieve this, you do not need more than one level of user exceptions. <br><br><pre> <code class="java hljs"><span class="hljs-function"><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">public</span></span></span><span class="hljs-function"> </span><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">void</span></span></span><span class="hljs-function"> </span><span class="hljs-title"><span class="hljs-function"><span class="hljs-title">persistCustomer</span></span></span><span class="hljs-params"><span class="hljs-function"><span class="hljs-params">(Customer c)</span></span></span><span class="hljs-function"> </span></span>{ <span class="hljs-comment"><span class="hljs-comment">// persist a Customer } public void manageCustomer(Customer c) throws MyBusinessException { // manage a Customer throw new MyBusinessException(e, e.getCode()); } public void createCustomer(Customer c) throws MyBusinessException { // create a Customer manageCustomer(c); }</span></span></code> </pre> </div><p>Source: <a href="https://habr.com/ru/post/337536/">https://habr.com/ru/post/337536/</a></p>
<section class="more-articles-navigation-panel js-more-articles-navigation-panel">
<h4>More articles:</h4>
<nav class="list-of-articles-container js-list-of-articles-container"><ul class="list-of-pages js-list-of-pages">
<li><a href="../337522/index.html">For a thousand rubles for the idea: we pay for proposals for the automation of QIWI wallets</a></li>
<li><a href="../337526/index.html">Microservice Architecture - Motive Shift to Target</a></li>
<li><a href="../337528/index.html">Parallelism versus multithreading versus asynchronous programming: an explanation</a></li>
<li><a href="../337532/index.html">10 years with hummingbirds: how it was</a></li>
<li><a href="../337534/index.html">Fads Stream API</a></li>
<li><a href="../337538/index.html">‚ÄúYoung people want to buy shares of familiar start-ups on the stock exchange‚Äù: finance trends according to the founder of the Robinhood service</a></li>
<li><a href="../337540/index.html">Setting up a Webpack 3 + Angular 4 development environment: from complex to simple</a></li>
<li><a href="../337546/index.html">Why don't CRM work</a></li>
<li><a href="../337548/index.html">Kaggle: how our nets considered sea lions in the Aleutian Islands</a></li>
<li><a href="../337550/index.html">Learn OpenGL. Part 2.4. - Texture Cards</a></li>
</ul></nav>
</section><br />
<a href="../../allArticles.html"><strong>All Articles</strong></a>
<script src="../../js/main.js"></script>

<!-- Yandex.Metrika counter -->
<script type="text/javascript" >
  (function (d, w, c) {
      (w[c] = w[c] || []).push(function() {
          try {
              w.yaCounter52496797 = new Ya.Metrika({
                  id:52496797,
                  clickmap:true,
                  trackLinks:true,
                  accurateTrackBounce:true,
                  webvisor:true
              });
          } catch(e) { }
      });

      var n = d.getElementsByTagName("script")[0],
          s = d.createElement("script"),
          f = function () { n.parentNode.insertBefore(s, n); };
      s.type = "text/javascript";
      s.async = true;
      s.src = "https://mc.yandex.ru/metrika/watch.js";

      if (w.opera == "[object Opera]") {
          d.addEventListener("DOMContentLoaded", f, false);
      } else { f(); }
  })(document, window, "yandex_metrika_callbacks");
</script>
<noscript><div><img src="https://mc.yandex.ru/watch/52496797" style="position:absolute; left:-9999px;" alt="" /></div></noscript>

<!-- Google Analytics -->
  <script>
    window.ga = function () { ga.q.push(arguments) }; ga.q = []; ga.l = +new Date;
    ga('create', 'UA-134931760-1', 'auto'); ga('send', 'pageview')
  </script>
  <script src="https://www.google-analytics.com/analytics.js" async defer></script>

</section>

  <footer class="page-footer">
    <div class="page-footer-legal-info-container page-footer-element">
      <p>
        Weekly-Geekly | <span class="page-footer-legal-info-year js-page-footer-legal-info-year">2019</span>
      </p>
    </div>
    <div class="page-footer-counters-container page-footer-element">
      <a class="page-footer-counter-clustrmap" href='#'  title='Visit tracker'><img src='https://clustrmaps.com/map_v2.png?cl=698e5a&w=271&t=t&d=6iCFw7uJz0zcOaoxz5k5PcLCJUzv2WG8G5V8M3U6Rc4&co=3a3a3a&ct=ffffff'/></a>
    </div>
  </footer>
</body>

</html>